### PR TITLE
refactor: remove existing vpc endpoint configurations from base module and move into sub-module

### LINKED ISSUE #462: VPC endpoint service complications with availability zone restrictions
I ran into a problem attempting to provision an SES endpoint: the default is to use all private subnets but some are for unspecified reasons unsupported — in my case us-east-1a,e,f are listed as “Service not supported in this Availability Zone” in the console and so an `apply` fails:

```
Error: Error creating VPC Endpoint: InvalidParameter: The VPC endpoint service com.amazonaws.us-east-1.email-smtp does not support the availability zone of the subnet: subnet-0de46d7dcc06da2a7.
```

### LINKED ISSUE #582: Solution for "Error: multiple VPC Endpoint Services matched"
If you have been using this VPC module and started from 2.2.2021 started seeing the error: `Error: multiple VPC Endpoint Services matched; use additional constraints to reduce matches to a single VPC Endpoint Service`, here are the available solutions depending on the versions you use:

## Terraform 0.12, 0.13, 0.14:

1. Upgrade Terraform AWS provider to be `3.10.0` (see [changelog for the release](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md#3100-october-09-2020)) or newer:

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 3.10"
    }
  }
}
```

2. Upgrade version of the module to be `2.70.0` (or newer):

```terraform
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "2.70.0"

  # ...
}
```

3. Alternatively, if you can't upgrade Terraform AWS provider as described in the first point, you can use this module version `2.69.0`:

```terraform
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "2.69.0"

  # ...
}
```

## Terraform 0.11

You can upgrade your configuration to use `v1.73.0` to get the specified issue fixed on Terraform 0.11 ([commit](https://github.com/terraform-aws-modules/terraform-aws-vpc/commit/691737859e2aaf45e0a84760a87ee9829e0229e3)):

```terraform
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "1.73.0"

  # ...
}
```

## Notes

If you want to manage the VPC endpoint for S3 outside of this module (when using Terraform 0.11, for eg), you can set `enable_s3_endpoint = false` and create resources `aws_vpc_endpoint` externally.

## References:

#573

https://github.com/hashicorp/terraform-provider-aws/issues/17417

### LINKED ISSUE #603: Allow both interface and gateway S3 VPC endpoints
It's perfectly allowable to have both an interface and gateway S3 VPC endpoint configured but currently this module allows one or the other using the `s3_endpoint_type` parameter. #575 proposed making the existing S3 endpoint parameters solely for the gateway type and adding a new set of parameters for managing the interface type, but this was not merged.

See the AWS documentation here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/privatelink-interface-endpoints.html

### LINKED ISSUE #607: Allow optional creation of S3 endpoint association for database subnets
Currenly, VPC module allows to create S3 endpoint for private and public subnets (optionally). It would be great to add possibility to create such endpoint (route table association) for database subnets as well. This could be very useful for enabling S3 access from RDS via S3 endpoint like described here: https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.SaveIntoS3.html 
and there:
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Integrating.Authorizing.Network.html

I think it could be done by simple code like:
```resource "aws_vpc_endpoint_route_table_association" "database_s3" {
  count = var.create_vpc && var.enable_s3_endpoint && var.enable_database_s3_endpoint && length(var.database_subnets) > 0 && var.s3_endpoint_type == "Gateway" ? local.nat_gateway_count : 0

  vpc_endpoint_id = aws_vpc_endpoint.s3[0].id
  route_table_id  = element(aws_route_table.database.*.id, count.index)
}
```
in `vpc-endpoints.tf`. 

BR
Grzegorz

### LINKED ISSUE #631: rds-data endpoint support
Hello since this endpoint is really important for all people using aws aurora serverless, can you please add it ?
It's quite the same as the others

### LINKED ISSUE #633: Expose VPC endpoint private IP on aws_vpc_endpoint resource
## Is your request related to a problem? Please describe.
I want to dynamically create security group egress rules that exclusively permit access to the CIDR of a VPC endpoint. However, the CIDR of the VPC endpoint is not exposed on the Terraform resource: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint#argument-reference 

As a result, I end up creating egress rules that just whitelist all traffic in my VPC, simply to allow them to hit the (unknown to Terraform) internal IP of the VPC endpoint.

## Describe the solution you'd like.
Just expose the CIDR of the resource, pls :) 

## Describe alternatives you've considered.
Hardcoding the IP as a variable :( 

## Additional context
This would probably require modifying the way the associated subnets of the resource are exposed, since an IP is assigned for each subnet the endpoint is deployed in.

### LINKED ISSUE #634: VPC Endpoints Configuration Does not Support "Non-Complete" Sets of VPCe Subnets
## Description

Hi, there. 

Recently, I faced an issue with lambda vpc endpoints not being created in my installation, as AWS simply did not provide those endpoints in all AZs. Current code in vpc-endpoints.tf presumes that we can simply re-use all the private subnets, while in fact this is not always the case. Please consult the documentation https://aws.amazon.com/ru/premiumsupport/knowledge-center/interface-endpoint-availability-zone/
which says that one can get an error of "Service not supported in this Availability Zone."
The solution would be to filter private subnets by the criterion of them being supported by the corresponding VPCe.

## ⚠️ Note

Before you submit an issue, please perform the following first:

1. Remove the local `.terraform` directory (! ONLY if state is stored remotely, which hopefully you are following that best practice!): `rm -rf .terraform/`
2. Re-initialize the project root to pull down modules: `terraform init`
3. Re-attempt your terraform plan or apply and check if the issue still persists

## Versions

 - Terraform:
 - any

 - Provider(s):
 - aws

 - Module:
 - terrraform-aws-vpc

## Reproduction

no

yes

Be lucky enough to have your VPCe not supported in all the AZs.
### Code Snippet to Reproduce

in my case it is "enable_lambda_endpoint" = true with empty lambda_endpoint_subnet_ids
## Expected behavior
The module should in fact filter the list of private subnets for endpoint availabilitiy by leveraging aws_vpc_endpoint_service data source.

## Actual behavior
Terraform fails with AWS Error "Service not supported in this Availability Zone."

### LINKED ISSUE #638: Support Amazon Redshift VPC endpoints
## Is your request related to a new offering from AWS?
AWS supports two different Redshift Interface VPC endpoints:

image

## Describe the solution you'd like.
This module should support deploying these two Redshift interface endpoints.

## Additional context
https://docs.aws.amazon.com/redshift/latest/mgmt/security-private-link.html

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- remove existing vpc endpoint configurations from base module and move into sub-module
- BREAKING: VPC endpoint resources are removed in favor of sub-module and `for_each` creation of any VPC endpoints
- BREAKING: AWS provider has been updated from v2.70 to v3.15 for VPC endpoint changes and available attributes
- combine issue/example projects into one project

## Motivation and Context
- due to VPC endpoint services being added all the time, and the ever growing complexity of the module, this change is to make the module more flexible and to allow users to choose what VPC endpoints they elect to use and attach to whichever resources they choose

Closes #454
Closes #462
Closes #560
Closes #582
Closes #603
Closes #604
Closes #607
Closes #608
Closes #611
Closes #613
Closes #618
Closes #626 
Closes #630
Closes #631
Closes #633
Closes #634
Closes #638
Closes #640

## Breaking Changes
Yes:
- BREAKING: VPC endpoint resources are removed in favor of sub-module and `for_each` creation of any VPC endpoints
- BREAKING: AWS provider has been updated from v2.70 to v3.15 for VPC endpoint changes and available attributes

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- I have updated and validated using the `complete-vpc` example

### CODE REVIEW

File: UPGRADE-3.0.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+If you find a bug, please open an issue with supporting configuration to reproduce.
+
+## List of backwards incompatible changes
+
+Previously, VPC endpoints were configured as standalone resources with their own set of variables and attributes. Now, this functionality is provided via a module which loops over a map of maps using `for_each` to generate the desired VPC endpoints. Therefore, to maintain the existing set of functionality while upgrading, you will need to perform the following changes:
+
+1. Move the endpoint resource from the main module to the sub-module. The example state move below is valid for all endpoints you might have configured (reference [`complete-vpc`](https://github.com/terraform-aws-modules/terraform-aws-vpc/tree/master/examples/complete-vpc) example for reference), where `ssmmessages` should be updated for and state move performed for each endpoint configured:
+
+```
+tf state mv 'module.vpc.aws_vpc_endpoint.ssm[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["ssm"]'
```
Comment (antonbabenko) [inline UPGRADE-3.0.md:16]: `tf` => `terraform`

File: UPGRADE-3.0.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   - `*_endpoint_security_group_ids`
+   - `*_endpoint_subnet_ids`
+   - `*_endpoint_private_dns_enabled`
+   - `*_endpoint_policy`
+
+2. Renamed variables:
+
+See the [VPC endpoint sub-module](modules/vpc-endpoints) for the more information on the variables to utilize for VPC endpoints
+
+   - None
```
Comment (antonbabenko) [inline UPGRADE-3.0.md:46]:
```suggestion
```

File: UPGRADE-3.0.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+3. Removed outputs:
+
+   - `vpc_endpoint_*`
+
+4. Renamed outputs:
+
+VPC endpoint outputs are now provided via the VPC endpoint sub-module and can be accessed via lookups. See [`complete-vpc`](https://github.com/terraform-aws-modules/terraform-aws-vpc/tree/master/examples/complete-vpc) for further examples of how to access VPC endpoint attributes from outputs.
+
+   - None
```
Comment (antonbabenko) [inline UPGRADE-3.0.md:56]:
```suggestion
```

File: examples/complete-vpc/main.tf [hunk 4]
```diff
@@ -1,20 +1,87 @@
...
+  name   = "complete-example"
   region = "eu-west-1"
+  tags = {
+    Owner       = "user"
+    Environment = "staging"
+    Name        = "complete"
+  }
 }
 
+################################################################################
```
Comment (antonbabenko) [inline examples/complete-vpc/main.tf:15]: Please move all "supporting resources" somewhere below the main `module "vpc"` block

File: modules/vpc-endpoints/README.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+See [`examples`](./examples) directory for working examples to reference:
+
+```hcl
+module "endpoints" {
+  source = "terraform-aws-modules/vpc/aws//vpc-endpoints"
+
+  vpc_id             = "vpc-12345678"
+  security_group_ids = ["sg-12345678"]
+
+  gateway_endpoints = {
```
Comment (antonbabenko) [inline modules/vpc-endpoints/README.md:20]: This is now combined into one `endpoints`.

File: modules/vpc-endpoints/README.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+- [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint)
+
+## Usage
+
+See [`examples`](./examples) directory for working examples to reference:
+
+```hcl
+module "endpoints" {
+  source = "terraform-aws-modules/vpc/aws//vpc-endpoints"
```
Comment (antonbabenko) [inline modules/vpc-endpoints/README.md:15]:
```suggestion
  source = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
```

File: modules/vpc-endpoints/variables.tf [hunk 7]
```diff
@@ -0,0 +1,34 @@
+variable "vpc_id" {
+  description = "The ID of the VPC in which the endpoint will be used"
+  type        = string
```
Comment (antonbabenko) [inline modules/vpc-endpoints/variables.tf:3]: Let's add `create` boolean flag to be able to disable this module the same way as we do in the VPC. For this reason there should be `default` in every variable.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Looks very good. Large cleanup. I commented just a few small things.

File: UPGRADE-3.0.md (hunk 1 above)
Comment (bryantbiggs) [reply UPGRADE-3.0.md]: updated in https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/635/commits/5c6a2313e49d8f537b8d0e4f45f827c62d742541

File: UPGRADE-3.0.md (hunk 2 above)
Comment (bryantbiggs) [reply UPGRADE-3.0.md]: updated in https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/635/commits/5c6a2313e49d8f537b8d0e4f45f827c62d742541

File: UPGRADE-3.0.md (hunk 3 above)
Comment (bryantbiggs) [reply UPGRADE-3.0.md]: updated in https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/635/commits/5c6a2313e49d8f537b8d0e4f45f827c62d742541

File: examples/complete-vpc/main.tf (hunk 4 above)
Comment (bryantbiggs) [reply examples/complete-vpc/main.tf:15]: updated in https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/635/commits/5c6a2313e49d8f537b8d0e4f45f827c62d742541

File: modules/vpc-endpoints/README.md (hunk 6 above)
Comment (bryantbiggs) [reply modules/vpc-endpoints/README.md]: good catch - updated in https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/635/commits/5c6a2313e49d8f537b8d0e4f45f827c62d742541

File: modules/vpc-endpoints/variables.tf (hunk 7 above)
Comment (antonbabenko) [reply modules/vpc-endpoints/variables.tf:9]: This is one of my favorite error messages :) objects vs objects battle...

Try writing like:

```hcl
endpoints = var.create ? tomap(var.endpoints) : tomap({})
```

Ref: https://github.com/terraform-aws-modules/terraform-aws-route53/blob/master/modules/zones/main.tf#L2

Comment (antonbabenko) [reply modules/vpc-endpoints/variables.tf:9]: Yes, this is the best way. :)

**Comment by bryantbiggs** [comment]
@antonbabenko friendly reminder 😉

File: examples/complete-vpc/main.tf [hunk 8]
```diff
@@ -1,20 +1,28 @@
 provider "aws" {
-  region = "eu-west-1"
+  region = local.region
```
Comment (antonbabenko) [inline examples/complete-vpc/main.tf]: I just hit this issue when using Terraform 0.15.0 - https://github.com/hashicorp/terraform-provider-aws/issues/8040#issuecomment-822774525 

Works well when I put a value there as a string:
```
region = "eu-west-1"
```

**Review by antonbabenko — APPROVED** [review]

File: UPGRADE-3.0.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+terraform state mv 'module.vpc.aws_vpc_endpoint.ssm[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["ssm"]'
+terraform state mv 'module.vpc.aws_vpc_endpoint.ssmmessages[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["ssmmessages"]'
+terraform state mv 'module.vpc.aws_vpc_endpoint.ec2[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["ec2"]'
+...
+```
+
+2. Remove the gateway endpoint route table association separate resources. The route table associations are now managed in the VPC endpoint resource itself via the map of maps provided to the VPC endpoint sub-module. Perform the necessary removals for each route table association and for S3 and/or DynamoDB depending on your configuration:
+
+```
+terraform state rm 'module.vpc.aws_vpc_endpoint_route_table_association.intra_dynamodb[0]'
```
Comment (antonbabenko) [inline UPGRADE-3.0.md:25]: I also verified all these `terraform state` commands. Works as expected.

**Comment by antonbabenko** [comment]
[v3.0.0](https://github.com/terraform-aws-modules/terraform-aws-vpc/releases/v3.0.0) has been just released. 🎉 🎉 🎉 🎉 🎉 🎉

**Comment by tabletcorry** [comment]
So if I am reading this correctly, the migration from 2.x to 3.x requires manual admin intervention to modify state for every use of this module, with the potential that a mis-typed `state mv` or `state rm` could damage production VPCs?

Or if we don't use state commands, we take a traffic double-blip on all endpoints as they are destroyed and then re-created.

This is a _very_ disruptive breaking change. I hope the 2.x line will continue to get updates if they required.

**Comment by antonbabenko** [comment]
Version 2.x won't be getting updates. We support just one major release at a time.

The process of upgrading from 2.x to 3.x indeed requires manual runs of `terraform state mv` commands if you can't allow recreation of resources (most of the production loads do not allow this).

I am pretty confident that [Upgrade Notes](https://github.com/terraform-aws-modules/terraform-aws-vpc/blob/master/UPGRADE-3.0.md) are correct, too.

**Comment by tabletcorry** [comment]
Ok, thank you for confirming.

I do want to express that this is quite painful for a large system where infrastructure-as-code has been implemented strictly, and virtually no one can run manual state commands. I realize there was some pain in maintaining this in the module, but the fix has pushed that pain onto every invocation of this module by every consumer.

**Comment by bryantbiggs** [comment]
hi @tabletcorry - we do our best to avoid breaking changes but as with all software projects, you reach a point where you have to make a significant change to continue making forward progress and best supporting the needs of those who use the module(s). the prior versions are all there for use and can be pinned (recommended) to avoid any breaking updates, but ultimately, it is up to each individual if they want to take on the challenge in updating in order to receive the latest changes. we do feel that this change was a long time coming - AWS releases VPC endpoints all the time and the prior solution was not scalable nor was it flexible enough for most folks (based on issues and PRs that we received). again, pinning the module version is going to be the safest route to avoid any disruptions until folks are ready to make the necessary changes to update (it is an opt in, its not something that is forced upon anyone)

**Comment by Routhinator** [comment]
How does `ses_endpoint_security_group_ids` get translated here? SES is missing from the "Complete" example and I don't see any way to attach security groups to the endpoints?

Is there any migration guide? How do we map the state shifts?

**Comment by Routhinator** [comment]
So, not sure how we migrate individual security groups for endpoints still..

However for others:

SES Example:
```
    ses = {
      service        = "email-smtp"
      tags           = { Name = "ses-vpc-endpoint" }
      subnet_ids     = module.vpc.private_subnets
    },
```

Migration of existing endpoints:
```
terraform state mv 'module.vpc.aws_vpc_endpoint.ses[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["ses"]'
terraform state mv 'module.vpc.aws_vpc_endpoint.s3[0]' 'module.vpc_endpoints.aws_vpc_endpoint.this["s3"]'
```
I only had these two endpoints to migrate, however it should be very similar for all of them.


I _cannot_ figure out how to provide both `module.vpc.private_subnets` and `module.vpc.database_subnets` to `subnet_ids` though.. Every permutation of concat and flatten I could think of failed to make an array it would accept. In my case I need to provide the DB subnet access to the endpoint for backups.

**Comment by acdha** [comment]
In addition to ses / email-smtp, I ran into “efs” to “elasticfilesystem”. It seems like it might be useful to have a list of some of the more common ones since `aws ec2 describe-vpc-endpoint-services --no-paginate --query "ServiceNames[*]"` isn't super intuitive.

**Comment by bryantbiggs** [comment]
@Routhinator you cannot associate the VPC endpoint with more than one subnet in the same AZ and why I suspect your permutations are failing. if you are creating a VPCe for SES, its one association in each AZ (typically via the private subnets)

**Pull request merged**